File: REST/eoq-discount/src/main/java/local/example/outcome/OutcomeResource.java
package local.example.outcome;

import local.example.outcome.model.Eoq;

import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.Set;

@Path("/outcome")
public class OutcomeResource {

    private final Set<Eoq> eoqs = Collections.newSetFromMap(
            Collections.synchronizedMap(new LinkedHashMap<>())
    );

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public Response read() {
        if (this.eoqs.isEmpty())
            return Response.ok().build();
        return Response.ok(this.eoqs).build();
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public Set<Eoq> compute(Eoq eoq) {
        eoq.setEoq();
        this.eoqs.add(eoq);
        return this.eoqs;
    }
}

File: REST/eoq-safety-rc3/src/main/kotlin/local/example/outcome/OutcomeResource.kt
package local.example.outcome

import local.example.outcome.model.Eoq
import java.util.*
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("/outcome")
class OutcomeResource {

    private val eoqs = Collections.newSetFromMap(
        Collections.synchronizedMap(
            LinkedHashMap<Eoq, Boolean>()
        )
    )

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    fun read(): Response? {
        return if (
            eoqs.isEmpty()
        ) Response.ok().build() else Response.ok(eoqs).build()
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    fun compute(eoq: Eoq): Set<Eoq?>? {
        eoq.setEoq()
        eoqs.add(eoq)
        return eoqs
    }
}
